compiler/rustc_trait_selection/src/solve/trait_goals.rs

```diff
@@ -599,10 +599,12 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
         // which will ICE for region vars.
         let args = ecx.tcx().erase_regions(goal.predicate.trait_ref.args);
 
-        let Some(assume) =
-            rustc_transmute::Assume::from_const(ecx.tcx(), goal.param_env, args.const_at(3))
-        else {
-            return Err(NoSolution);
+        let maybe_assume =
+            rustc_transmute::Assume::from_const(ecx.tcx(), goal.param_env, args.const_at(3));
+        let assume = match maybe_assume {
+            Some(Ok(assume)) => assume,
+            Some(Err(_guar)) => return Err(NoSolution),
+            None => return Err(NoSolution),
```

Comment:
same here

Reply:
Since `NoSolution` was just a struct, I created a new enum `QueryError` so I could return `QueryError::NoSolution` or `QueryError::Ambiguous`. But the change might be too invasive. Let me know what you think! I made the change in a separate commit.

Reply:
Thanks, that makes a lot more sense! `evaluate_added_goals_and_make_query_response` doesn't exist, did you mean `evaluate_added_goals_and_make_canonical_response`?